### PR TITLE
Feature/dynamic elements

### PR DESCRIPTION
**Pull request by twltwl**
Related issue https://github.com/zeit/styled-jsx/issues/147

Based on comment by @giuseppeg 

Automaticly scopes elements if possible ( works if declared inside of component )

for other cases a styled-jsx prop can be added

```javascript
// works
const Cmp = () => {
  const Element = 'div';
  return (<Element className="foo">
    <style jsx>{`.foo{...}`}</style>
  </Element>)
}

// wont work
const Element = 'div';
const Cmp = () => (<Element className="foo">
  <style jsx>{`.foo{...}`}</style>
</Element>)

// works
const Element = 'div';
const Cmp = () => (<Element className="foo" styled-jsx>
  <style jsx>{`.foo{...}`}</style>
</Element>)
```

### CODE REVIEW

File: src/babel.js [hunk 1]
```diff
@@ -40,11 +41,21 @@ export default function({ types: t }) {
...
         if (
           name &&
           name !== 'style' &&
           name !== STYLE_COMPONENT &&
-          name.charAt(0) !== name.charAt(0).toUpperCase()
+          (name.charAt(0) !== name.charAt(0).toUpperCase() ||
+            Object.values(path.scope.bindings).some(binding =>
+              binding.referencePaths.some(r => r === tag)
+            ) ||
+            (attributes &&
```
Comment (giuseppeg) [inline src/babel.js]: can we remove this feature for now?

File: test/fixtures/attribute-generation-classname-rewriting.js [hunk 2]
```diff
@@ -34,6 +36,9 @@ export default () => (
     <div {...props} {...rest} />
     <div {...props} data-foo {...rest} />
     <div {...props} className={'test'} data-foo {...rest} />
+    <Element styled-jsx />
```
Comment (giuseppeg) [inline test/fixtures/attribute-generation-classname-rewriting.js]: I won't want to include this feature here for now, let's fix one thing at the time

File: test/fixtures/dynamic-element.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    <Element className="root">
+      <p>dynamic element</p>
+      <style jsx>{`
+        .root {
+          background: red;
+        }
+      `}</style>
+    </Element>
+  )
+}
```
Comment (giuseppeg) [inline test/fixtures/dynamic-element.js]: awesome can you add a few more tests cases? i.e. a few more combinations to where `Element` is defined outside of the component, and maybe when it is lowercase too (any edge case you can think of). Also can you add a class component too?

File: src/babel.js (hunk 1 above)
Comment (twltwl) [reply src/babel.js]: Sure, if we want to add it at a later point let me know

File: test/fixtures/dynamic-element.js (hunk 3 above)
Comment (twltwl) [reply test/fixtures/dynamic-element.js]: Maybe Im missing something, but I dont think it will work if Element is defined outside of the component. ( Thats why I added the styled-jsx option )
I´ll add the testcases

Comment (giuseppeg) [reply test/fixtures/dynamic-element.js]: correct, it won't work (it won't add the className) and we want a snapshot of that! ;)

Comment (twltwl) [reply test/fixtures/dynamic-element.js]: 👍

**Comment by giuseppeg** [comment]
@twltwl this branch is good to merge, can you just remove node `6` from travis https://github.com/zeit/styled-jsx/blob/a55c821a998b9a7210c03a30c4ae33de7ec94eb6/.travis.yml#L3 ? and maybe add a new line for node `10`?

**Comment by giuseppeg** [comment]
@twltwl thanks so much for this great contribution! 🙌

**Comment by iamawebgeek** [comment]
@giuseppeg I cannot use this feature for some reason. Babel is not stripping out the `styled-jsx` attribute and adding a scoped `className`. How can I figure out what is the problem?
